File: shopee-scraper/browser/python/shopee.py
"""Shopee scraper using the official Scrapeless Python SDK + Playwright over CDP.

Shopee hydrates PDP fields client-side from an XHR to ``/api/v4/pdp/get_pn``.
The response holds ``data.item`` (title, price fields, images, stock, rating
summary), ``data.product_price``, ``data.shop_detailed`` and a breadcrumb under
``data.product_category``. We capture that XHR live; the SSR HTML is too sparse
to rely on.
"""

from __future__ import annotations

import asyncio
import json
import os
import re
from dataclasses import asdict, dataclass, field
from typing import Any, Dict, List, Optional

from parsel import Selector
from playwright.async_api import Response, async_playwright
from scrapeless import Scrapeless
from scrapeless.types import ICreateBrowser

DEFAULT_PROXY_COUNTRY = "TH"
DEFAULT_SESSION_TTL = 240
DEFAULT_HOST = "https://shopee.co.th"


@dataclass
class Product:
    id: str
    url: str
    title: str
    brand: Optional[str] = None
    price: Optional[str] = None
    originalPrice: Optional[str] = None
    discount: Optional[str] = None
    currency: Optional[str] = None
    rating: Optional[float] = None
    reviews: Optional[int] = None
    images: List[str] = field(default_factory=list)
    seller: Optional[str] = None
    sellerUrl: Optional[str] = None
    availability: Optional[str] = None
    description: Optional[str] = None
    categories: List[str] = field(default_factory=list)


@dataclass
class SearchResult:
    id: str
    title: str
    url: str
    image: Optional[str] = None
    price: Optional[str] = None
    originalPrice: Optional[str] = None
    discount: Optional[str] = None
    rating: Optional[float] = None
    reviews: Optional[int] = None
    location: Optional[str] = None


def _client() -> Scrapeless:
    if not os.environ.get("SCRAPELESS_API_KEY") and not os.environ.get("SCRAPELESS_KEY"):
        raise RuntimeError(
            "SCRAPELESS_API_KEY env var not set. Sign up at https://app.scrapeless.com"
        )
    if not os.environ.get("SCRAPELESS_API_KEY") and os.environ.get("SCRAPELESS_KEY"):
        os.environ["SCRAPELESS_API_KEY"] = os.environ["SCRAPELESS_KEY"]
    return Scrapeless()


def _text(s: Optional[str]) -> Optional[str]:
    if not s:
        return None
    cleaned = re.sub(r"\s+", " ", s).strip()
    return cleaned or None


def _to_float(s) -> Optional[float]:
    if s is None:
        return None
    if isinstance(s, (int, float)):
        return float(s)
    m = re.sub(r"[^0-9.]", "", str(s))
    if not m:
        return None
    try:
        return float(m)
    except ValueError:
        return None


def _to_int(s) -> Optional[int]:
    if s is None:
        return None
    if isinstance(s, (int, float)):
        return int(s)
    m = re.sub(r"[^0-9]", "", str(s))
    if not m:
        return None
    try:
        return int(m)
    except ValueError:
        return None


def _abs(url: Optional[str], host: str = DEFAULT_HOST) -> Optional[str]:
    if not url:
        return None
    if url.startswith("//"):
        return "https:" + url
    if url.startswith("http"):
        return url
    return host + (url if url.startswith("/") else "/" + url)


def _id_from_url(url: Optional[str]) -> str:
    if not url:
        return ""
    m = re.search(r"i\.(\d+)\.(\d+)", url)
    if m:
        return m.group(2)
    m2 = re.search(r"-i\.(\d+)\.(\d+)", url)
    if m2:
        return m2.group(2)
    return ""


def _img(image_id: Optional[str]) -> Optional[str]:
    if not image_id:
        return None
    if image_id.startswith("http"):
        return image_id
    return f"https://down-th.img.susercontent.com/file/{image_id}"


def _price(value) -> Optional[str]:
    # Shopee encodes prices as integers scaled by 100000.
    if value is None:
        return None
    try:
        n = float(value) / 100000.0
    except (TypeError, ValueError):
        return None
    if n <= 0:
        return None
    return f"฿{n:,.2f}"


def _uniq(seq):
    out, seen = [], set()
    for x in seq:
        if not x or x in seen:
            continue
        seen.add(x)
        out.append(x)
    return out


def parse_product_from_pdp(data: Dict[str, Any], url: str) -> Product:
    item = data.get("item") or {}
    product_price = data.get("product_price") or {}
    shop = data.get("shop_detailed") or data.get("shop") or {}
    breadcrumb = data.get("product_category") or data.get("breadcrumb") or []

    price_block = product_price.get("price") or {}
    single = price_block.get("single_value")
    before = (product_price.get("before_discount") or {}).get("single_value")
    rebate = product_price.get("rebate_percentage")

    images: List[str] = []
    for iid in item.get("images") or []:
        u = _img(iid)
        if u:
            images.append(u)
    images = _uniq(images)

    brand = None
    pb = item.get("brand")
    if isinstance(pb, dict):
        brand = pb.get("name")
    elif isinstance(pb, str):
        brand = pb

    categories = [
        b.get("display_name") or b.get("name")
        for b in (breadcrumb or [])
        if isinstance(b, dict) and (b.get("display_name") or b.get("name"))
    ]

    rating_summary = item.get("item_rating") or {}
    rating_val = rating_summary.get("rating_star")
    reviews_val = None
    rt = rating_summary.get("rating_count")
    if isinstance(rt, list) and rt:
        reviews_val = rt[0]
    elif isinstance(rt, (int, float)):
        reviews_val = rt

    stock = item.get("stock")
    availability = None
    if isinstance(stock, (int, float)):
        availability = "In Stock" if stock > 0 else "Out of Stock"

    shop_id = (item.get("shopid") or shop.get("shopid"))
    seller_url = _abs(f"/shop/{shop_id}") if shop_id else None

    return Product(
        id=str(_id_from_url(url) or item.get("itemid") or ""),
        url=url,
        title=item.get("title") or item.get("name") or "",
        brand=brand,
        price=_price(single),
        originalPrice=_price(before),
        discount=(f"-{rebate}%" if isinstance(rebate, (int, float)) and rebate else None),
        currency=product_price.get("currency") or item.get("currency"),
        rating=_to_float(rating_val),
        reviews=_to_int(reviews_val),
        images=images,
        seller=shop.get("name") or shop.get("account", {}).get("username") if isinstance(shop, dict) else None,
        sellerUrl=seller_url,
        availability=availability,
        description=item.get("description"),
        categories=categories,
    )


def parse_search(html: str) -> List[SearchResult]:
    sel = Selector(text=html)
    out: List[SearchResult] = []
    seen: set[str] = set()
    cards = [
        "li[data-sqe='item']",
        "div[data-sqe='item']",
        "div[class*='shopee-search-item-result__item']",
        "div[class*='col-xs-2-4']",
    ]
    for css in cards:
        for card in sel.css(css):
            href = card.css("a[href*='-i.']::attr(href), a[href*='i.']::attr(href)").get() or ""
            if href.startswith("//"):
                href = "https:" + href
            if href.startswith("/"):
                href = DEFAULT_HOST + href
            if not href:
                continue
            item_id = _id_from_url(href)
            if not item_id or item_id in seen:
                continue
            seen.add(item_id)
            title = _text(card.css("[class*='line-clamp']::text, [class*='name']::text, div[class*='_10Wbs-']::text").get()) or card.css("a::attr(title)").get() or _text(card.css("a::text").get()) or ""
            image = card.css("img::attr(src)").get() or card.css("img::attr(data-src)").get()
            if image and image.startswith("//"):
                image = "https:" + image
            price = _text(card.css("[class*='price']::text, span[class*='ZEgDH9']::text").get())
            original_price = _text(card.css("[class*='origin']::text, [class*='line-through']::text").get())
            discount = _text(card.css("[class*='discount']::text, [class*='percent']::text").get())
            rating = _to_float(_text(card.css("[class*='rating']::text, [class*='shopee-rating-stars__lit']::attr(style)").get()))
            if rating is not None and rating > 5:
                rating = rating / 20
            if rating is not None and rating > 5:
                rating = None
            reviews = _to_int(_text(card.css("[class*='sold']::text, [class*='review']::text").get()))
            location = _text(card.css("[class*='location']::text, [class*='shopee-search-item-result__item'] [class*='ZkPYTL']::text").get())
            out.append(SearchResult(
                id=item_id,
                title=title,
                url=href,
                image=image,
                price=price,
                originalPrice=original_price,
                discount=discount,
                rating=rating,
                reviews=reviews,
                location=location,
            ))
    return out


async def scrape_product(product_url: str) -> Product:
    client = _client()
    session = client.browser.create(ICreateBrowser(proxy_country=DEFAULT_PROXY_COUNTRY, session_ttl=DEFAULT_SESSION_TTL))
    async with async_playwright() as pw:
        browser = await pw.chromium.connect_over_cdp(session.browser_ws_endpoint)
        try:
            page = await browser.new_page()
            holder: Dict[str, Any] = {"body": None}

            async def _on_response(resp: Response) -> None:
                try:
                    if holder["body"]:
                        return
                    u = resp.url.lower()
                    if "/api/v4/pdp/get_pn" not in u and "/api/v4/item/get" not in u:
                        return
                    holder["body"] = await resp.text()
                except Exception:
                    pass

            page.on("response", lambda r: asyncio.create_task(_on_response(r)))
            await page.goto(product_url, wait_until="domcontentloaded", timeout=60000)
            for _ in range(60):
                if holder["body"]:
                    break
                await page.wait_for_timeout(500)
            if not holder["body"]:
                raise RuntimeError(
                    f"shopee: detail XHR (/api/v4/pdp/get_pn) never fired for {product_url}"
                )
            env = json.loads(holder["body"])
            data = (env or {}).get("data") or {}
            if not data.get("item"):
                raise RuntimeError("shopee: detail XHR payload missing item block")
            return parse_product_from_pdp(data, product_url)
        finally:
            try:
                await browser.close()
            except Exception:
                pass


async def scrape_search(query: str, max_pages: int = 1, host: str = DEFAULT_HOST) -> List[SearchResult]:
    client = _client()
    session = client.browser.create(ICreateBrowser(proxy_country=DEFAULT_PROXY_COUNTRY, session_ttl=DEFAULT_SESSION_TTL))
    results: List[SearchResult] = []
    async with async_playwright() as pw:
        browser = await pw.chromium.connect_over_cdp(session.browser_ws_endpoint)
        try:
            page = await browser.new_page()
            for p in range(0, max_pages):
                url = f"{host}/search?keyword={query.replace(' ', '%20')}&page={p}"
                await page.goto(url, wait_until="domcontentloaded", timeout=60000)
                try:
                    await page.wait_for_selector("li[data-sqe='item'], div[data-sqe='item']", timeout=20000)
                except Exception:
                    pass
                try:
                    await page.evaluate(
                        "() => new Promise(r => { let y = 0; const i = setInterval(() => { window.scrollBy(0, 600); y += 600; if (y >= document.body.scrollHeight) { clearInterval(i); r(); } }, 120); })"
                    )
                    await page.wait_for_timeout(1500)
                except Exception:
                    pass
                html = await page.content()
                results.extend(parse_search(html))
        finally:
            try:
                await browser.close()
            except Exception:
                pass
    return results


def to_dict(obj: Any) -> Any:
    if hasattr(obj, "__dataclass_fields__"):
        return asdict(obj)
    if isinstance(obj, list):
        return [to_dict(x) for x in obj]
    if isinstance(obj, dict):
        return {k: to_dict(v) for k, v in obj.items()}
    return obj

File: shopee-scraper/browser/nodejs/shopee.mjs
// Shopee scraper using the official @scrapeless-ai/sdk + puppeteer-core over CDP.
//
// Shopee hydrates PDP fields client-side from an XHR to
//   /api/v4/pdp/get_pn
// The response holds `data.item` (title, images, stock, rating summary),
// `data.product_price`, `data.shop_detailed` and a breadcrumb under
// `data.product_category`. We capture that XHR live; the SSR HTML is too sparse
// to rely on.

import { Scrapeless } from "@scrapeless-ai/sdk";
import puppeteer from "puppeteer-core";
import * as cheerio from "cheerio";

const DEFAULT_PROXY_COUNTRY = "TH";
const DEFAULT_SESSION_TTL = 240;
const DEFAULT_HOST = "https://shopee.co.th";

function requireKey() {
  const k = process.env.SCRAPELESS_API_KEY ?? process.env.SCRAPELESS_KEY;
  if (!k) throw new Error("SCRAPELESS_API_KEY env var not set. Sign up at https://app.scrapeless.com");
  return k;
}

function client() {
  return new Scrapeless({ apiKey: requireKey() });
}

async function withBrowser(fn, { proxyCountry = DEFAULT_PROXY_COUNTRY, retries = 2, label = "navigation" } = {}) {
  let lastError;
  for (let attempt = 0; attempt <= retries; attempt++) {
    const { browserWSEndpoint } = await client().browser.create({ proxyCountry, sessionTTL: DEFAULT_SESSION_TTL });
    let browser;
    try {
      browser = await puppeteer.connect({ browserWSEndpoint });
      return await fn(browser);
    } catch (e) {
      lastError = e;
      if (attempt === retries) break;
      await new Promise((r) => setTimeout(r, 4000 * Math.pow(1.5, attempt)));
    } finally {
      try { await browser?.close(); } catch (_) {}
    }
  }
  throw new Error(`${label}: giving up after ${retries + 1} attempts: ${lastError?.message ?? lastError}`);
}

// ---------------- helpers ----------------

function abs(url, host = DEFAULT_HOST) {
  if (!url) return null;
  if (url.startsWith("//")) return "https:" + url;
  if (url.startsWith("http")) return url;
  return host + (url.startsWith("/") ? url : "/" + url);
}

function uniq(arr) {
  return Array.from(new Set(arr.filter(Boolean)));
}

function idFromUrl(url) {
  if (!url) return "";
  const m = url.match(/i\.(\d+)\.(\d+)/);
  if (m) return m[2];
  return "";
}

function img(imageId) {
  if (!imageId) return null;
  if (imageId.startsWith("http")) return imageId;
  return `https://down-th.img.susercontent.com/file/${imageId}`;
}

// Shopee encodes prices as integers scaled by 100000.
function price(value) {
  if (value === null || value === undefined) return null;
  const n = Number(value) / 100000;
  if (!Number.isFinite(n) || n <= 0) return null;
  return `฿${n.toLocaleString("en-US", { minimumFractionDigits: 2, maximumFractionDigits: 2 })}`;
}

function toFloat(text) {
  if (text === null || text === undefined) return null;
  if (typeof text === "number") return Number.isFinite(text) ? text : null;
  const m = String(text).replace(/[^0-9.]/g, "");
  if (!m) return null;
  const n = parseFloat(m);
  return Number.isNaN(n) ? null : n;
}

function toInt(text) {
  if (text === null || text === undefined) return null;
  if (typeof text === "number") return Number.isFinite(text) ? Math.trunc(text) : null;
  const m = String(text).replace(/[^0-9]/g, "");
  if (!m) return null;
  const n = parseInt(m, 10);
  return Number.isNaN(n) ? null : n;
}

// ---------------- parsers ----------------

// Build a `Product` dict from the pdp `data` blob.
export function parseProductFromPdp(data, url) {
  const item = data.item || {};
  const productPrice = data.product_price || {};
  const shop = data.shop_detailed || data.shop || {};
  const breadcrumb = data.product_category || data.breadcrumb || [];

  const priceBlock = productPrice.price || {};
  const single = priceBlock.single_value;
  const before = (productPrice.before_discount || {}).single_value;
  const rebate = productPrice.rebate_percentage;

  const imgs = [];
  for (const iid of item.images || []) {
    const u = img(iid);
    if (u) imgs.push(u);
  }
  const images = uniq(imgs);

  const categories = (breadcrumb || [])
    .map((b) => b?.display_name || b?.name)
    .filter(Boolean);

  const ratingSummary = item.item_rating || {};
  const ratingVal = ratingSummary.rating_star;
  let reviewsVal = null;
  const rt = ratingSummary.rating_count;
  if (Array.isArray(rt) && rt.length) reviewsVal = rt[0];
  else if (typeof rt === "number") reviewsVal = rt;

  let availability = null;
  if (typeof item.stock === "number") availability = item.stock > 0 ? "In Stock" : "Out of Stock";

  const shopId = item.shopid || shop.shopid;

  return {
    id: String(idFromUrl(url) || item.itemid || ""),
    url,
    title: item.title || item.name || "",
    brand: (item.brand && (item.brand.name || item.brand)) || null,
    price: price(single),
    originalPrice: price(before),
    discount: typeof rebate === "number" && rebate ? `-${rebate}%` : null,
    currency: productPrice.currency || item.currency || null,
    rating: toFloat(ratingVal),
    reviews: toInt(reviewsVal),
    images,
    seller: shop.name || shop.account?.username || null,
    sellerUrl: shopId ? abs(`/shop/${shopId}`) : null,
    availability,
    description: item.description || null,
    categories,
  };
}

export function parseSearch(html) {
  const $ = cheerio.load(html);
  const out = [];
  const seen = new Set();

  const cards = [
    "li[data-sqe='item']",
    "div[data-sqe='item']",
    "div[class*='shopee-search-item-result__item']",
    "div[class*='col-xs-2-4']",
  ];
  for (const sel of cards) {
    $(sel).each((_, el) => {
      const card = $(el);
      const a = card.find("a[href*='i.']").first();
      let href = a.attr("href") || "";
      if (href.startsWith("//")) href = "https:" + href;
      if (href.startsWith("/")) href = DEFAULT_HOST + href;
      if (!href) return;
      const id = idFromUrl(href);
      if (!id || seen.has(id)) return;
      seen.add(id);

      const text$ = (n) => (n?.text?.() || "").replace(/\s+/g, " ").trim() || null;
      const title =
        text$(card.find("[class*='line-clamp'], [class*='name'], div[class*='_10Wbs-']").first()) ||
        a.attr("title") ||
        text$(a) ||
        "";

      const image =
        card.find("img").first().attr("src") ||
        card.find("img").first().attr("data-src") ||
        null;
      const priceText = text$(card.find("[class*='price'], span[class*='ZEgDH9']").first()) || null;
      const originalPrice = text$(card.find("[class*='origin'], [class*='line-through']").first()) || null;
      const discount = text$(card.find("[class*='discount'], [class*='percent']").first()) || null;
      let rating = toFloat(
        card.find("[class*='shopee-rating-stars__lit']").first().attr("style") ||
          text$(card.find("[class*='rating']").first()),
      );
      if (rating !== null && rating > 5) rating = rating / 20;
      if (rating !== null && rating > 5) rating = null;
      const reviews = toInt(text$(card.find("[class*='sold'], [class*='review']").first()));
      const location = text$(card.find("[class*='location'], [class*='ZkPYTL']").first()) || null;

      out.push({
        id,
        title,
        url: href,
        image: image && image.startsWith("//") ? "https:" + image : image,
        price: priceText,
        originalPrice,
        discount,
        rating,
        reviews,
        location,
      });
    });
  }
  return out;
}

// ---------------- scrape functions ----------------

export async function scrapeProduct(productUrl) {
  return withBrowser(async (browser) => {
    const page = await browser.newPage();
    const bodies = [];
    page.on("response", async (resp) => {
      try {
        const u = resp.url().toLowerCase();
        if (!u.includes("/api/v4/pdp/get_pn") && !u.includes("/api/v4/item/get")) return;
        const txt = await resp.text();
        if (txt) bodies.push(txt);
      } catch (_) {}
    });
    await page.goto(productUrl, { waitUntil: "domcontentloaded", timeout: 60000 });
    // The detail page makes 1–2 pdp calls during boot. Wait for one carrying
    // `item`, with up to 35s budget.
    const deadline = Date.now() + 35000;
    let data = null;
    while (Date.now() < deadline) {
      for (const body of bodies) {
        try {
          const env = JSON.parse(body);
          const d = env?.data || null;
          if (d && d.item) { data = d; break; }
        } catch (_) {}
      }
      if (data) break;
      await new Promise((r) => setTimeout(r, 500));
    }
    if (!data) {
      throw new Error(`shopee: detail XHR (/api/v4/pdp/get_pn) never returned an item block for ${productUrl}`);
    }
    return parseProductFromPdp(data, productUrl);
  }, { label: "scrape_product" });
}

export async function scrapeSearch(query, maxPages = 1, host = DEFAULT_HOST) {
  return withBrowser(async (browser) => {
    const page = await browser.newPage();
    const out = [];
    for (let p = 0; p < maxPages; p++) {
      const url = `${host}/search?keyword=${encodeURIComponent(query)}&page=${p}`;
      await page.goto(url, { waitUntil: "domcontentloaded", timeout: 60000 });
      try { await page.waitForSelector("li[data-sqe='item'], div[data-sqe='item']", { timeout: 20000 }); } catch (_) {}
      try {
        await page.evaluate(async () => {
          await new Promise((r) => {
            let y = 0;
            const i = setInterval(() => {
              window.scrollBy(0, 600);
              y += 600;
              if (y >= document.body.scrollHeight) { clearInterval(i); r(); }
            }, 120);
          });
        });
      } catch (_) {}
      await new Promise((r) => setTimeout(r, 1500));
      const html = await page.content();
      out.push(...parseSearch(html));
    }
    return out;
  }, { label: "scrape_search" });
}
